prow/cmd/exporter/collector_test.go

```diff
@@ -79,6 +88,219 @@ func assertEqual(t *testing.T, actual, expected interface{}) {
 	}
 }
 
+type fakeLister struct {
+}
+
+func (l fakeLister) List(selector labels.Selector) ([]*prowapi.ProwJob, error) {
+	return []*prowapi.ProwJob{
+		{
+			Spec: prowapi.ProwJobSpec{
+				Agent: prowapi.KubernetesAgent,
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: "default",
+				Name:      "7785d7a6-e601-11e9-8512-da8015665453",
+				Labels: map[string]string{
+					"created-by-prow":          "true",
+					"event-GUID":               "770bab40-e601-11e9-8e50-08c45d902b6f",
+					"preset-bazel-scratch-dir": "true",
+					"preset-service-account":   "true",
+					"prow.k8s.io/job":          "pull-test-infra-bazel",
+					"prow.k8s.io/refs.org":     "kubernetes",
+					"prow.k8s.io/refs.pull":    "14543",
+					"prow.k8s.io/refs.repo":    "test-infra",
+					"prow.k8s.io/type":         "presubmit",
+				},
+				Annotations: map[string]string{
+					"prow.k8s.io/job":            "pull-test-infra-bazel",
+					"testgrid-create-test-group": "true",
+				},
+			},
+		},
+		{
+			Spec: prowapi.ProwJobSpec{
+				Agent: prowapi.KubernetesAgent,
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: "default",
+				Name:      "e44f91e5-e604-11e9-99c1-0a58ac10f9a6",
+				Labels: map[string]string{
+					"created-by-prow":       "true",
+					"event-GUID":            "e4216820-e604-11e9-8cf0-295472589b4f",
+					"prow.k8s.io/job":       "branch-ci-openshift-release-master-config-updates",
+					"prow.k8s.io/refs.org":  "openshift",
+					"prow.k8s.io/refs.repo": "release",
+					"prow.k8s.io/type":      "postsubmit",
+				},
+				Annotations: map[string]string{
+					"prow.k8s.io/job": "branch-ci-openshift-release-master-config-updates",
+				},
+			},
+		},
+	}, nil
+}
+
+func TestProwJobCollector(t *testing.T) {
+
+	pjc := prowJobCollector{
+		lister: fakeLister{},
+	}
+	c := make(chan prometheus.Metric)
+	go pjc.Collect(c)
+
+	counter := 0
+	var labelMetric1, annotationMetric1, labelMetric2, annotationMetric2 prometheus.Metric
+	waiting := true
+
+	for {
+		if !waiting {
+			break
+		}
+		select {
+		case msg := <-c:
+			counter++
+			logrus.WithField("counter", counter).Infof("received a metric")
+			switch counter {
+			case 1:
+				labelMetric1 = msg
+			case 2:
+				annotationMetric1 = msg
+			case 3:
+				labelMetric2 = msg
+			case 4:
+				annotationMetric2 = msg
+			}
+			if counter > 4 {
+				logrus.WithField("counter", counter).Errorf("should not have received more than 4 metrics, existing ...")
+				waiting = false
+			}
+		case <-time.After(time.Second):
+			waiting = false
```

Comment:
Nit: A simpler approach would be to just call `t.Fatal` here, then you don't need the `waiting` variable

Reply:
Probably i missed your point here.
Even though in the case that we have received 4 metrics in the channel, we still need timeout to exit the for loop.
In this case, timeout just means we are done waiting for metrics sent to the channel. Timeout does not imply failure of the case, right?

```
$ go test -v -run TestProwJobCollector ./prow/cmd/exporter/...
go: finding github.com/mattbaird/jsonpatch latest
=== RUN   TestProwJobCollector
time="2019-10-07T09:18:28-04:00" level=info msg="received a metric" counter=1
time="2019-10-07T09:18:28-04:00" level=info msg="received a metric" counter=2
time="2019-10-07T09:18:28-04:00" level=info msg="received a metric" counter=3
time="2019-10-07T09:18:28-04:00" level=info msg="received a metric" counter=4
time="2019-10-07T09:18:29-04:00" level=info msg="timeout, existing ..." counter=4
time="2019-10-07T09:18:29-04:00" level=info msg="get all 4 metrics"
--- PASS: TestProwJobCollector (1.00s)
PASS
ok      k8s.io/test-infra/prow/cmd/exporter     (cached)

```

Reply:
Once you have received four metrics (counter == 4) you can just `close` the channel (so subsequent writes to it will cause a panic and make the tests fail) and `break` the loop.

Also it isn't really needed to have four distinct variables for the metrics, you can just use use a `var receivedMetrics []prometheus.Metric` and then `append` to it until `counter == 4`.

All of this is nitpicking thought and I am fine merging it without, if you want.

Reply:
> Once you have received four metrics (counter == 4) you can just `close` the channel (so subsequent writes to it will cause a panic and make the tests fail) and `break` the loop.
> 

Now i see your point. Updated the code.
This would get rid of the var. `wait` with `break` AND the `goto` as we need to break [the outer](https://stackoverflow.com/a/51996662) `for` loop (suggesting a better way than `goto`?).

> Also it isn't really needed to have four distinct variables for the metrics, you can just use use a `var receivedMetrics []prometheus.Metric` and then `append` to it until `counter == 4`.
> 

Done.

> All of this is nitpicking thought and I am fine merging it without, if you want.

I knew it marked it as `nit` but i will try to make it if I understand it is improving the code. Thank you very much for that.